cpp/src/arrow/ipc/memory.cc

```diff
@@ -0,0 +1,152 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+#include "arrow/ipc/memory.h"
+
+#include <sys/mman.h> // For memory-mapping
+#include <algorithm>
+#include <cstdint>
+#include <cstdio>
+#include <cstring>
+#include <string>
+
+#include "arrow/util/buffer.h"
+#include "arrow/util/status.h"
+
+namespace arrow {
+namespace ipc {
+
+MemorySource::MemorySource(AccessMode access_mode) :
+    access_mode_(access_mode) {}
+
+MemorySource::~MemorySource() {}
+
+// Implement MemoryMappedSource
+
+class MemoryMappedSource::Impl {
+ public:
+  Impl() :
+      file_(nullptr),
+      is_open_(false),
+      data_(nullptr) {}
+
+  ~Impl() {
+    if (is_open_) {
+      munmap(data_, size_);
+      fclose(file_);
+    }
+  }
+
+  Status Open(const std::string& path, MemorySource::AccessMode mode) {
+    path_ = path;
+
+    if (mode == MemorySource::READ_WRITE) {
+      file_ = fopen(path.c_str(), "r+b");
+    } else {
+      file_ = fopen(path.c_str(), "rb");
+    }
+    if (file_ == nullptr) {
+      return Status::IOError("Unable to open file");
+    }
+
+    fseek(file_, 0L, SEEK_END);
+    if (ferror(file_)) {
+      return Status::IOError("Unable to seek to end of file");
```

Comment:
After a little googling, this is outside my domain of knowledge how to do this in a platform-independent way. let's leave it to a downstream JIRA